File: cpp/src/file/fragment.cpp


#include "milvus-storage/file/fragment.h"
#include <memory>
#include <vector>
#include "assert.h"
#include "proto/manifest.pb.h"

namespace milvus_storage {
Fragment::Fragment(std::int64_t fragment_id) : fragment_id_(fragment_id) {}

void Fragment::add_file(const std::string& file) { files_.push_back(file); }

const std::vector<std::string>& Fragment::files() const { return files_; }

std::int64_t Fragment::id() const { return fragment_id_; }

void Fragment::set_id(int64_t id) { fragment_id_ = id; }

std::unique_ptr<manifest_proto::Fragment> Fragment::ToProtobuf() const {
  auto fragment_proto = std::make_unique<manifest_proto::Fragment>();
  for (const auto& file : files_) {
    fragment_proto->add_files(file);
  }
  fragment_proto->set_id(fragment_id_);
  return fragment_proto;
}

std::unique_ptr<Fragment> Fragment::FromProtobuf(const manifest_proto::Fragment& fragment) {
  auto res = std::make_unique<Fragment>(fragment.id());
  for (const auto& file : fragment.files()) {
    res->add_file(file);
  }
  return res;
}
}  // namespace milvus_storage

File: go/file/fragment/fragment.go
package fragment

import "github.com/milvus-io/milvus-storage/go/proto/manifest_proto"

type FragmentType int32

const (
	kUnknown FragmentType = 0
	kData    FragmentType = 1
	kDelete  FragmentType = 2
)

type Fragment struct {
	fragmentId int64
	files      []string
}

type FragmentVector []Fragment

func ToFilesVector(fragments []Fragment) []string {
	files := make([]string, 0)
	for _, fragment := range fragments {
		files = append(files, fragment.files...)
	}
	return files
}

func NewFragment(fragmentId int64) *Fragment {
	// TODO: check fragmentId
	return &Fragment{
		fragmentId: fragmentId,
		files:      make([]string, 0),
	}
}

func (f *Fragment) AddFile(file string) {
	f.files = append(f.files, file)
}

func (f *Fragment) Files() []string {
	return f.files
}

func (f *Fragment) FragmentId() int64 {
	return f.fragmentId
}

func (f *Fragment) SetFragmentId(fragmentId int64) {
	f.fragmentId = fragmentId
}

func (f *Fragment) ToProtobuf() *manifest_proto.Fragment {
	fragment := &manifest_proto.Fragment{}
	fragment.Id = f.fragmentId
	for _, file := range f.files {
		fragment.Files = append(fragment.Files, file)
	}
	return fragment
}

func FromProtobuf(fragment *manifest_proto.Fragment) *Fragment {
	newFragment := NewFragment(fragment.Id)
	for _, file := range fragment.Files {
		newFragment.files = append(newFragment.files, file)
	}
	return newFragment
}
